### PR TITLE
Initial support for thumbnails for existing files.

### LINKED ISSUE #58: Thumbnail previews of media files available on Wikimedia Commons
Scenario: user wants to use OpenRefine to add structured data to existing files from Wikimedia Commons. They load a series of file paths from Wikimedia Commons and reconcile them with Wikimedia Commons.

We want to give users the option to toggle previews (thumbnails) of the media files, and the option to click the thumbnails to enlarge them. Such preview thumbnails are helpful during editing (e.g. to check if a certain thing is indeed depicted in a file, without having to click on to the file page on Commons).

Wireframes for this feature have been drawn by @lozanaross and OpenRefine/OpenRefine#5154 provides the technical basis for making this feature possible.

Most recent wireframes I found (v4, development version):
* The small thumbnails: https://xd.adobe.com/view/df767737-c4ca-4d25-977e-7ad8ca3a6840-8a25/screen/8c10c8d4-9ab6-4da2-adb6-9552f789593f/
* Hover effect: https://xd.adobe.com/view/df767737-c4ca-4d25-977e-7ad8ca3a6840-8a25/screen/e8aadeae-d6e6-414f-9852-26e85ce6964d/
* Enlarged thumbnail: https://xd.adobe.com/view/df767737-c4ca-4d25-977e-7ad8ca3a6840-8a25/screen/698c6f1c-d747-4adc-895b-142e21c2830c/
image

### PR DESCRIPTION
**Pull request by wetneb**
Closes #58.

This changes requires to run OpenRefine from the `wetneb:5154-extension-point-for-cell-rendering` branch (which corresponds to the PR https://github.com/OpenRefine/OpenRefine/pull/5312).
The aim is that this PR is merged before 3.7, so that the extension can rely on it to implement the thumbnails.

Changes to the thumbnails themselves can then be made independently of OpenRefine's release cycle.

Since this PR is only about existing files, I did not add an explicit toggle to turn thumbnails on and off - they are just enabled by default for all cells that are reconciled to a media file on a Wikibase instance.

Thumbnails are only generated for file types which have thumbnailing enabled in MediaWiki (roughly, images and videos).

One thing I would probably still add is a waiting spinner when the full screen image is being loaded. It would also be worth checking the thumbnails are requested in a way that maximizes the chances of cache hits (which I am not sure I got right so far).

## Screenshot

image

### CODE REVIEW

**Comment by trnstlntk** [comment]
I tested it on a variety of files, and it works excellently. It indeed has that wow factor! I get thumbnails for images, videos and PDFs. In some cases (I saw it with a PDF and a video) the pop-up loads slowly indeed (is that due to file size?) and a spinner would indeed make sense to indicate that something is happening.

The size of the thumbnails is great IMO. Maybe other users will have different opinions but I think it's a good size to be able to discern files well without overwhelming.

I can imagine that a toggle function will be useful to add later, but this is absolutely great for now.

File: module/scripts/project/thumbnail-renderer.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        .appendTo(div);
+    div.on('click', function(evt) {
+       div.remove();
+    });
+  }
+
+  /*
+   Important: thumbnails take resources to generate, so it is worth sticking to the thumbnail sizes
+   which are auto-generated at upload time:
+   320, 640, 800, 1024, 1280, 1920, 2560, 2880
```
Comment (trnstlntk) [inline module/scripts/project/thumbnail-renderer.js:118]: Interesting, makes sense. Which one did you use now, 640?

**Review by trnstlntk — APPROVED** [review]

**Review by lozanaross — APPROVED** [review]

**Comment by trnstlntk** [comment]
@wetneb I noticed that this PR hasn't been merged "back in the days" (I can't recall why?) - but thumbnails do work in the Commons extension, except for this little annoying bug here: #89 

Can you recall why this would still be open?

Thanks for digging into your memory :-)

**Comment by trnstlntk** [comment]
> @wetneb I noticed that this PR hasn't been merged "back in the days" (I can't recall why?) - but thumbnails do work in the Commons extension, except for this little annoying bug here: #89
> 
> Can you recall why this would still be open?
> 
> Thanks for digging into your memory :-)

If you prefer not to dig into this anymore (which is fine), I'll take the freedom to close this PR as not done, to reduce confusion.

**Comment by wetneb** [comment]
I think you do need to merge this PR - I do not think thumbnail support works without it. Could it be that the version of the Commons extension you are using is generated from this PR?

So I would recommend merging this and releasing a new version of the extension.

**Comment by trnstlntk** [comment]
> I think you do need to merge this PR - I do not think thumbnail support works without it. Could it be that the version of the Commons extension you are using is generated from this PR?
> 
> So I would recommend merging this and releasing a new version of the extension.

Thanks, you're entirely correct! I now feel very stupid 🙈 - I was indeed working from this PR myself.

I notice we released 0.1.0 without thumbnail support (it was also not included in the release notes). I prefer for someone to continue adjusting this existing PR to fix the #89 bug, and also someone else to release the new version. I did the previous release myself but found it very difficult, I had absolutely no idea what I was doing except for following instructions blindly. IMO it's a job for someone with stronger technical skills and familiarity with Git/GitHub processes.

I'll see how I can adjust the bounty to include these requirements.

**Review by trnstlntk — CHANGES_REQUESTED** [review]
After quite a bit more testing of this PR, I have noticed that it doesn't support all file extensions on Wikimedia Commons, see issue #89 (capitalized file names)

I suggest to update this PR accordingly, then merge, and do a new release of the extension.

**Comment by vitaly-zdanevich** [comment]
But I have previews only on mouse hover:
image

**Pull request merged**